### PR TITLE
Eng 5990 partial index

### PR DESCRIPTION
**Pull request by jxchinaren**

### CODE REVIEW

File: tests/frontend/org/voltdb/planner/TestIndexSelection.java [hunk 1]
```diff
@@ -527,4 +580,17 @@ private void checkIndexPredicateDoesNotHave(AbstractPlanNode pn, String[] column
...
+    private void checkIndexSkipNullPredicateIsNull(AbstractPlanNode pn, boolean isTrue) {
+        assertEquals(1, pn.getChildCount());
+        pn = pn.getChild(0);
+        String json = pn.toJSONString();
+        if (pn instanceof IndexCountPlanNode) {
+            assertTrue(isTrue == (!json.contains("SKIP_NULL_PREDICATE")));
+        } else {
+            // index scan
+            AbstractExpression skipNull = ((IndexScanPlanNode) pn).getSkipNullPredicate();
+            assertTrue(isTrue == (skipNull == null));
```
Comment (jxchinaren) [inline tests/frontend/org/voltdb/planner/TestIndexSelection.java]: @alexeevm Better with "assertEquals(isTrue,  skipNull == null);"

Same applies to Line:589

File: tests/frontend/org/voltdb/planner/TestIndexSelection.java [hunk 2]
```diff
@@ -527,4 +580,17 @@ private void checkIndexPredicateDoesNotHave(AbstractPlanNode pn, String[] column
             }
         }
     }
+
+    private void checkIndexSkipNullPredicateIsNull(AbstractPlanNode pn, boolean isTrue) {
```
Comment (jxchinaren) [inline tests/frontend/org/voltdb/planner/TestIndexSelection.java]: @alexeevm Better to rename "isTrue" to be "hasSkipNullPredicate". 
Rename "checkIndexSkipNullPredicateIsNull" to be "checkIndexSkipNullPredicate".

Then 
assertEquals(hasSkipNullPredicate,  json.contains("SKIP_NULL_PREDICATE") );
assertEquals(hasSkipNullPredicate, skipNull != null);

File: tests/frontend/org/voltdb/planner/TestIndexSelection.java [hunk 3]
```diff
@@ -444,41 +445,93 @@ public void testPartialIndexComparisonPredicateNonExactMatch()
...
+            checkIndexName(pn, PlanNodeType.INDEXCOUNT, "\"TARGET_INDEX_NAME\":\"PARTIAL_IDX_6\"");
+            checkIndexSkipNullPredicateIsNull(pn, true);
+        }
+        {
+            //CREATE INDEX partial_idx_6 ON c (g) where g is not null;
+            // skipNull predicate is redundant and eliminated
+            AbstractPlanNode pn = compile("select e from c where g > 0;");
+            checkIndexName(pn, PlanNodeType.INDEXSCAN, "\"TARGET_INDEX_NAME\":\"PARTIAL_IDX_6\"");
+            checkIndexSkipNullPredicateIsNull(pn, true);
+        }
```
Comment (jxchinaren) [inline tests/frontend/org/voltdb/planner/TestIndexSelection.java]: @alexeevm Can you have another partial index like "CREATE INDEX partial_idx_7 ON c (g) where g > 0;" and test the same above two queries to check whether the IS NULL Predicate is added or not?

File: src/frontend/org/voltdb/plannodes/IndexScanPlanNode.java [hunk 4]
```diff
@@ -159,18 +163,24 @@ public void setSkipNullPredicate() {
     }
 
     public void setSkipNullPredicate(int nextKeyIndex) {
+        m_skip_null_predicate = buildSkipNullPredicate(nextKeyIndex, m_catalogIndex, m_tableScan, m_searchkeyExpressions);
```
Comment (jxchinaren) [inline src/frontend/org/voltdb/plannodes/IndexScanPlanNode.java]: @alexeevm Nice refactoring.

**Comment by jxchinaren** [comment]
Looks good. Merge to the branch now for testing.

File: tests/frontend/org/voltdb/planner/TestIndexSelection.java (hunk 3 above)
Comment (alexeevm) [inline tests/frontend/org/voltdb/planner/TestIndexSelection.java]: Yes. Added.
CREATE INDEX partial_idx_6 ON c (g) where g < 0;
CREATE INDEX partial_idx_7 ON c (g) where g is not null;
select count(*) from c where g < 0; - Index partial_idx_6, skipNullPredicate is NULL
select * from c where g < 0; - Index partial_idx_6, skipNullPredicate is NULL

File: tests/frontend/org/voltdb/planner/TestIndexSelection.java (hunk 2 above)
Comment (alexeevm) [inline tests/frontend/org/voltdb/planner/TestIndexSelection.java]: Yes, indeed. Renamed

File: tests/frontend/org/voltdb/planner/TestIndexSelection.java (hunk 1 above)
Comment (alexeevm) [inline tests/frontend/org/voltdb/planner/TestIndexSelection.java]: Corrected

**Pull request merged**